bpf/bpf_lxc.c

```diff
@@ -824,6 +825,26 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
 skip_egress_gateway:
 #endif
 
+#ifdef ENABLE_VTEP
+	{
+		int i;
+
+		for (i = 0; i < VTEP_NUMS; i++) {
+			if (tunnel_endpoint == VTEP_ENDPOINT[i]) {
+				vtep_mac = VTEP_MAC[i];
+				break;
+			}
+		}
+
+		if (vtep_mac && tunnel_endpoint) {
+			if (eth_store_daddr(ctx, (__u8 *)&vtep_mac, 0) < 0)
+				return DROP_WRITE_ERROR;
+			return __encap_and_redirect_with_nodeid(ctx, tunnel_endpoint,
+									WORLD_ID, monitor);
```

Comment:
I noticed one subtle thing with this function:

Previously, the second-to-last `WORLD_ID` parameter here would always be the source pod's identity, so this function does two things:
* Set the VNI to this value
* Send a notification to userspace using this identity as the source identity (basically pass the same variable to `send_trace_notify`).

So an example `cilium monitor` output would look like this when sending traffic between pods in the network:

```
-> overlay flow 0x0 , identity 19929->unknown state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.244.0.253 -> 10.244.1.18 EchoRequest                                                                                                                                                                                                                                        
```

Note `identity 19929->unknown` basically says that the local identity is `19929` and the remote identity is unknown. This implies that the pod's identity is `19929` which we could cross-reference with `cilium identity list` to understand.

Now when enabling this feature and sending traffic to the VTEP, we see flows like this in `cilium monitor`:

```
-> overlay flow 0x0 , identity world->unknown state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.244.0.253 -> 10.1.5.1 EchoRequest                                                                                                                                                                                                                                           
```

Now we shouldn't interpret the `identity world->unknown` as "the pod's identity is `world` and it's sending to unknown". Instead we should intepret a flow like this to say "we're sending a packet to the overlay with source identity world", ie VNI 2. :thinking: 

I suspect that this may confuse Hubble output at some point, so we may want to consider adding another new parameter so that this call can pass two arguments: (1) VNI and (2) src_identity. All other callers of this function would pass `seclabel` into both parameters. This one would pass `WORLD_ID` into the vni parameter and pass `seclabel` into the src_identity parameter. Then inside, it would use `src_identity` to log the flow via `send_trace_notify`, but use `vni` to actually populate the packet.

Either way, I don't think that this is a big problem for a beta version, just a slightly odd visibility glitch.

Reply:
thanks for the observation, I will keep this on my mind after we resolved the network policy support issue of this feature https://github.com/cilium/cilium/issues/17694, the network policy support is my first priority to move this feature to stable version :-)

Reply:
created https://github.com/cilium/cilium/issues/18459 to track the note here for my future reference to address the issue